// @vueuse/firebase/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vueDemi = require('vue-demi');
var shared = require('@vueuse/shared');

function useAuth(authInstance) {
    let auth;
    if (authInstance instanceof Function)
        auth = authInstance();
    else
        auth = authInstance;
    const user = vueDemi.ref(auth.currentUser);
    const isAuthenticated = vueDemi.computed(() => !!user.value);
    auth.onIdTokenChanged(authUser => user.value = authUser);
    return {
        isAuthenticated,
        user,
    };
}

function getData(docRef) {
    const data = docRef.data();
    if (data) {
        Object.defineProperty(data, 'id', {
            value: docRef.id.toString(),
            writable: false,
        });
    }
    return data;
}
function isDocumentReference(docRef) {
    var _a;
    return (((_a = docRef.path) === null || _a === void 0 ? void 0 : _a.match(/\//g)) || []).length % 2 !== 0;
}
/**
 * Reactive Firestore binding. Making it straightforward to always keep your
 * local data in sync with remotes databases.
 *
 * @see https://vueuse.org/useFirestore
 * @param docRef
 * @param initialValue
 * @param options
 */
function useFirestore(docRef, initialValue = undefined, options = {}) {
    const { 
    // eslint-disable-next-line no-console
    errorHandler = (err) => console.error(err), autoDispose = true, } = options;
    if (isDocumentReference(docRef)) {
        const data = vueDemi.ref(initialValue);
        const close = docRef.onSnapshot((snapshot) => {
            data.value = getData(snapshot) || null;
        }, errorHandler);
        shared.tryOnUnmounted(() => {
            close();
        });
        return data;
    }
    else {
        const data = vueDemi.ref(initialValue);
        const close = docRef.onSnapshot((snapshot) => {
            data.value = snapshot.docs.map(getData).filter(shared.isDef);
        }, errorHandler);
        if (autoDispose) {
            shared.tryOnUnmounted(() => {
                close();
            });
        }
        return data;
    }
}

/**
 * Reactive Firebase Realtime Database binding.
 *
 * @param docRef
 * @param options
 */
function useRTDB(docRef, options = {}) {
    const { autoDispose = true, } = options;
    const data = vueDemi.ref(undefined);
    function update(snapshot) {
        data.value = snapshot.val();
    }
    docRef.on('value', update);
    if (autoDispose) {
        shared.tryOnUnmounted(() => {
            docRef.off('value', update);
        });
    }
    return data;
}

exports.useAuth = useAuth;
exports.useFirestore = useFirestore;
exports.useRTDB = useRTDB;
